modules/queue/queue_redis.go

```diff
@@ -0,0 +1,222 @@
+	closed    chan struct{}
+	exemplar  interface{}
+	workers   int
+	name      string
+}
+
+// RedisQueueConfiguration is the configuration for the redis queue
+type RedisQueueConfiguration struct {
+	Network      string
+	Addresses    string
+	Password     string
+	DBIndex      int
+	BatchLength  int
+	QueueLength  int
+	QueueName    string
+	Workers      int
+	MaxWorkers   int
+	BlockTimeout time.Duration
+	BoostTimeout time.Duration
+	BoostWorkers int
+	Name         string
+}
+
+// NewRedisQueue creates single redis or cluster redis queue
+func NewRedisQueue(handle HandlerFunc, cfg, exemplar interface{}) (Queue, error) {
+	configInterface, err := toConfig(RedisQueueConfiguration{}, cfg)
+	if err != nil {
+		return nil, err
+	}
+	config := configInterface.(RedisQueueConfiguration)
+
+	dbs := strings.Split(config.Addresses, ",")
+
+	dataChan := make(chan Data, config.QueueLength)
+	ctx, cancel := context.WithCancel(context.Background())
+
+	var queue = &RedisQueue{
+		pool: &WorkerPool{
+			baseCtx:            ctx,
+			cancel:             cancel,
+			batchLength:        config.BatchLength,
+			handle:             handle,
+			dataChan:           dataChan,
+			blockTimeout:       config.BlockTimeout,
+			boostTimeout:       config.BoostTimeout,
+			boostWorkers:       config.BoostWorkers,
+			maxNumberOfWorkers: config.MaxWorkers,
+		},
+		queueName: config.QueueName,
+		exemplar:  exemplar,
+		closed:    make(chan struct{}),
+		workers:   config.Workers,
+		name:      config.Name,
+	}
+	if len(dbs) == 0 {
+		return nil, errors.New("no redis host found")
+	} else if len(dbs) == 1 {
+		queue.client = redis.NewClient(&redis.Options{
+			Network:  config.Network,
+			Addr:     strings.TrimSpace(dbs[0]), // use default Addr
+			Password: config.Password,           // no password set
+			DB:       config.DBIndex,            // use default DB
+		})
+	} else {
+		queue.client = redis.NewClusterClient(&redis.ClusterOptions{
+			Addrs: dbs,
+		})
+	}
+	if err := queue.client.Ping().Err(); err != nil {
+		return nil, err
+	}
+	queue.pool.qid = GetManager().Add(queue, RedisQueueType, config, exemplar, queue.pool)
+
+	return queue, nil
+}
+
+// Run runs the redis queue
+func (r *RedisQueue) Run(atShutdown, atTerminate func(context.Context, func())) {
+	atShutdown(context.Background(), r.Shutdown)
+	atTerminate(context.Background(), r.Terminate)
+
+	go func() {
+		_ = r.pool.AddWorkers(r.workers, 0)
+	}()
+
+	go r.readToChan()
+
+	log.Trace("RedisQueue: %s Waiting til closed", r.name)
+	<-r.closed
+	log.Trace("RedisQueue: %s Waiting til done", r.name)
+	r.pool.Wait()
+
+	log.Trace("RedisQueue: %s Waiting til cleaned", r.name)
+	ctx, cancel := context.WithCancel(context.Background())
+	atTerminate(ctx, cancel)
+	r.pool.CleanUp(ctx)
+	cancel()
+}
+
+func (r *RedisQueue) readToChan() {
+	for {
+		select {
+		case <-r.closed:
+			// tell the pool to shutdown
+			r.pool.cancel()
+			return
+		default:
+			bs, err := r.client.LPop(r.queueName).Bytes()
```

Comment:
Won't this lock until `LPop` returns? It seems to me that the lock should be handled at the same level as `<-r.closed`; otherwise `<-r.closed` won't be checked until after anything is read from the client.

Reply:
So it depends on what you mean by blocking... 

* Redis LPop doesn't block at the Redis end - so once you've sent the message and redis has received it, it will send back redis.Nil.

* If you mean the queue blocking whilst this is going on well... It's difficult because LPop is not a channel so you can't do:
 
```go
case <- r.closed():
...
case <- LPop():
...
```

Looking at the code for go-redis (the current v6 line) - although it has functions `WithContext(...)` and `Context()` these aren't being used anywhere... They're a bit more plugged in in the v7 beta - so although I can add the WithContext() formulation to the code it won't be helpful. (I'm also a bit concerned that there's no clear way to just create a RedisClient with a context from the start but looking at that code the uncloned previous client doesn't seem to need to be closed.)

The default Timeout for read and write are 3 seconds - which I suppose are ok.

The only way of cancelling I can see is to create a purpose built dialer. I guess that's what we have to do if we really want this cancellable but then upgrading to v7 when it comes out will be a bit more difficult.

Reply:
OK so in order to have the context be in this connection section we would need to rewrite the line:

```go
	} else if len(dbs) == 1 {
		queue.client = redis.NewClient(&redis.Options{
			Network:  config.Network,
			Addr:     strings.TrimSpace(dbs[0]), // use default Addr
			Password: config.Password,           // no password set
			DB:       config.DBIndex,            // use default DB
		})
```

To:

```go
		addr := strings.TrimSpace(dbs[0])
		var dialer func() (net.Conn, error)
		if config.TLS {
			dialer = func() (net.Conn, error) {
				netDialer := &net.Dialer{
					Timeout:   5 * time.Second,
					KeepAlive: 5 * time.Minute,
				}
				return netDialer.DialContext(ctx, config.Network, addr)
			}
		} else {
			host := addr
			host, _, err := net.SplitHostPort(addr)
			if err != nil {
				log.Warn("Unable to split %s as a host port: Error: %v", addr)
				host = addr
			}
			tlsConfig := &tls.Config{ServerName: host}
			dialer = func() (net.Conn, error) {
				netDialer := &net.Dialer{
					Timeout:   5 * time.Second,
					KeepAlive: 5 * time.Minute,
				}
				errChan := make(chan error, 2)
				time.AfterFunc(netDialer.Timeout, func() {
					errChan <- fmt.Errorf("tls: DialWithDialer timed out")
				})
				conn, err := netDialer.DialContext(ctx, config.Network, addr)
				if err != nil {
					return nil, err
				}
				tlsConn := tls.Client(conn, tlsConfig)
				go func() {
					errChan <- tlsConn.Handshake()
				}()
				select {
				case err = <-errChan:
					if err != nil {
						conn.Close()
						return nil, err
					}
				case <-ctx.Done():
					tlsConn.Close()
					conn.Close()
					return nil, ctx.Err()
				}

				return tlsConn, nil
			}
		}
		queue.client = redis.NewClient(&redis.Options{
			Network:  config.Network,
			Addr:     addr,
			Password: config.Password,
			DB:       config.DBIndex,
			Dialer:   dialer,
		})
```

Reply:
And something similar for the cluster client...

Reply:
Otherwise the shutdown time will be increased in up to three seconds. Perhaps we can shorten the wait to 1 sec? (is it configurable?)

Reply:
it could be configurable as go-redis provides the option.